### PR TITLE
Drop broken code for fake ajax redirect.

### PR DESCRIPTION
**Pull request by blieb**
As discussed on irc with @hmic. Dropping the feature of fake ajax redirect. 

This breaks my app. ($this->redirect() does not when doing an ajax request)
It is using the deprecated RequestActionTrait. Fixing it is unnecessary base it will be removed soon.

### CODE REVIEW

File: src/Controller/Component/RequestHandlerComponent.php [hunk 1]
```diff
@@ -120,7 +120,6 @@ public function implementedEvents()
         return [
             'Controller.startup' => 'startup',
             'Controller.beforeRender' => 'beforeRender',
-            'Controller.beforeRedirect' => 'beforeRedirect',
```
Comment (dereuromark) [inline src/Controller/Component/RequestHandlerComponent.php]: For me those changes look rather BC breaking, at least the method must stay, so people extending will not suddenly have their beforeRedirect not executed anymore IMO.

Comment (blieb) [inline src/Controller/Component/RequestHandlerComponent.php]: Good point, placed that piece of code back.

**Comment by hmic** [comment]
+1

**Comment by dereuromark** [comment]
I am pretty sure those changes will break my https://github.com/dereuromark/cakephp-ajax/blob/master/src/Controller/Component/AjaxComponent.php#L121 - as so far it seemed to still work to render those pages in some cases just fine.
( Live example at http://sandbox.dereuromark.de/sandbox/ajax-examples/redirecting-prevented ).

Can you confirm this is still working afterwards?

**Comment by dereuromark** [comment]
Wait..
Earlier you write

> when I pass array to requestAction() it works ok, but when I pass an string it fails

So in fact for arrays it was and is working fine. You want to remove sth that works for everyone who uses the proper way of using URLs internally? And it only fails for you because you use the wrong way (String URLs)?

👎 You should throw an exception instead when String is being passed. That would be the reasonable solution here IMO

**Comment by hmic** [comment]
@dereuromark: wrong!
look at the code! it converts array urls into strings itself, then parses the string and passes the string to requestAction, which in turn fails, ever! he was talking about requestAction() not failing on proper array urls, this function fails in any case!

**Comment by hmic** [comment]
@lorenzo, @ADmad: merge it already! :p

**Comment by blieb** [comment]
Controller->redirect() creates a string from the array. 

When i look in the code of requestAction() it should work. String urls are handled. But it had a bug. And the code Will be removed in the future.

It is almost broken for a year now.

So if your was working before the merge. It Will work after the merge :)

**Comment by hmic** [comment]
Thanks for confirming @dereuromark!

**Comment by ADmad** [comment]
While I would love for this feature to go away we can simply remove the relevant code as it would be a BC break. Someone out there might be relying on this wacky feature.

@blieb How exactly is this causing you problems?

**Comment by blieb** [comment]
@admad it does not forward the params to the request. So it does not know what controller should be used. It will give an exception  that the controller with name "Controller" does not exist. 

It is broken for almost a year. I've seen a issue that is describing this problem in October 2015.

The code does not work at at and will. It will break apps that use it. So I there will be nog BC issues.

File: tests/TestCase/Controller/Component/RequestHandlerComponentTest.php [hunk 2]
```diff
@@ -899,217 +865,6 @@ public function testPrefers()
...
-        $this->Controller->RequestHandler->request = $this->Controller->request;
-        $this->Controller->RequestHandler->response = $this->Controller->response;
-        $this->Controller->request->expects($this->any())->method('is')->will($this->returnValue(true));
-
-        $response = $this->Controller->RequestHandler->beforeRedirect(
-            $event,
-            ['controller' => 'RequestHandlerTest', 'action' => 'destination'],
-            $this->Controller->response
-        );
-        $this->assertRegExp('/posts index/', $response->body(), 'RequestAction redirect failed.');
```
Comment (markstory) [inline tests/TestCase/Controller/Component/RequestHandlerComponentTest.php]: Doesn't this test contradict your statements about this code 'never working'?

File: tests/TestCase/Controller/Component/RequestHandlerComponentTest.php [hunk 3]
```diff
@@ -899,217 +865,6 @@ public function testPrefers()
...
-            ->with('ajax')
-            ->will($this->returnValue(true));
-
-        $response = $this->Controller->RequestHandler->beforeRedirect(
-            $event,
-            '/request_action/params_pass?a=b&x=y?ish',
-            $this->Controller->response
-        );
-        $data = json_decode($response, true);
-        $this->assertEquals('/request_action/params_pass', $data['here']);
```
Comment (markstory) [inline tests/TestCase/Controller/Component/RequestHandlerComponentTest.php]: Same goes for this one?

Comment (blieb) [inline tests/TestCase/Controller/Component/RequestHandlerComponentTest.php]: Looks like the tests are failing at this point. In the next request the controller name is unknown. 

The only difference I see is that the Request class does not receive $params['Params']['controller']; (and action). That is the point where it fails.

Comment (blieb) [inline tests/TestCase/Controller/Component/RequestHandlerComponentTest.php]: I've created a simple controller to test it. Just create a new cakephp project. and add this to TestController.php: [TestController.txt](https://github.com/cakephp/cakephp/files/489332/TestController.txt)


When done go to: http://localhost/test/testRedirect and you see it works fine.

now simulate an ajax request with curl:
curl -H "X-Requested-With: XMLHttpRequest" "http://localhost/test/testRedirect"

You will see an exception that the controller with name 'Controller' does not exist.

Comment (markstory) [inline tests/TestCase/Controller/Component/RequestHandlerComponentTest.php]: Ok I'll give that a shot, but I'm still not clear on how the requestAction requests are working in the tests. These tests are looking at the response generated by the requestAction controller action.

**Comment by markstory** [comment]
Closing as deleting the code is not the solution, I've put up a pull request that fixes the root cause of the issue.

**Comment by hmic** [comment]
We were talking about this "featrue" on IRC extensively.
@ADmad and @lorenzo despite other non-contributors wanted it to go away. It is, at best, black-magic that should be avoided.
Additionally, as it has been broken for some time now, even with the tests passing nobody *could* have used it lately, it sounds save to apply the 80/20 rule and mark it deprecated *now* and remove in 3.3.6, 3.4 the latest.

Please comment!

**Comment by ADmad** [comment]
> as it has been broken for some time now, even with the tests passing nobody could have used it lately

It was broken since 3.3.0 due to the changes related to middleware, it worked before that and does now too with the fixes @markstory made.

>  mark it deprecated now and remove in 3.3.6, 3.4 the latest.

Yes it should be marked deprecated but as per our BC guidelines it can't be removed until 4.0

**Comment by hmic** [comment]
I need to come back to this 3.3.0 middleware thing...
This f*** middleware introduction broke like half the projects out there and it has been supposed to be BC!
Next time somebody complains about not beeing able to do a trivial non-bc release upgrade, I know where to point them to! I am struggling ever other day with ppl that come to the chat and complain about there 2.5 apps and not beeing able to upgrade (still).
It has been rushed in instead of beeing a 4.0. Now you are debating about BC breaks in a feature that *obviously*, as it has been broken unnoticed already, nobody uses, on short notice?
I don't give much to the BC guidelines with how the middleware got rushed in, and so do all of you (core members). Don't try argue about it, it's in the history! It has not been meant this way for sure, still it was a mistake backed by the team.
As you can obviously see on my agressive attitude, I have been and still am angry about how you (the core members) state your decisions and opinions.
This is mine.

**Comment by ADmad** [comment]
Given all the issues with had with the introduction of middleware I do agree with you that perhaps it wasn't the best idea to do it in a minor release. Lesson learned. But that doesn't mean we should throw the BC guidelines out of the window. We will try to be more careful in future when introducing new features. Your opinions about middleware have been heard loud and clear but beating up a dead horse isn't productive. :smile: 

> I have been and still am angry about how you (the core members) state your decisions and opinions.

You are welcome to make suggestions on how we can improve upon that.

**Comment by hmic** [comment]
This lesson should have been learned on cake-2.6 already! This could have been a BC breaking 3.0 release, 3.0 could have been 4.0 and 3.3 - 5.0 already! This would have been "sane" to go by the BC guidelines.
It just failed another time!

**Comment by inoas** [comment]
**OT (Middleware)**

I can see reasons that CakePHP 3.3 could have been CakePHP 4.0 (say, version numbers are cheap).

However one great thing about CakePHP 2 has been that it has been supported for so very long.
Having a lot more major releases means you will have to pay the price of having no long term suppor.

Say 3.3 had been 4.0 then the currently named 4.0 (aka 5.0) would have probably dropped a lot support, e.g. for the current 3.0, 3.1, 3.2 and 3.3 without any choice to get into a branch that is maintained long term and not too many upgrade hassles. Take alone the docs. Would you have forked them for 4.0-based-on-3.3? 3.x would then have suffered a lot.

**TLDR:** So I am neither angry nor disappointed. I am not sure it was the best decision but I am neither sure that the other way around it would have been - long term - a lot better decision. 

I guess the best way to go forward is hear the people that have issues with Dispatch vs Middleware and patch things up. And as far as I follow the repo, exactly that is happening.

**Comment by hmic** [comment]
@ionas LTS versions solves that problems. Keep 2.x LTS, 3.x not LTS, but upgrading to 4.x is cheap...

**Comment by blieb** [comment]
I think that every javascript library just follow redirects. Correct me if i'm wrong. 

So removing this will not cause any BC problems. Happy it will be fixed in the next release, but it would also be nice if I was able to turn it off. I don't see any advantages of this feature.

**Comment by hmic** [comment]
you can turn it off: implement the event youself, return false to stop propagation. done.

**Comment by markstory** [comment]
@hmic I did my best at making the middleware changes backwards compatible, but due to gaps in my testing and the almost complete lack of help from the community a number of issues were missed. I apologize for that and I'm doing everything I can to fix the issues as quickly as they are reported. I hope that you can understand that, and that you are not required or forced to upgrade any release we put out.

To be completely honest your angry, unhelpful and dismissive attitude is discouraging. It makes me question why I work countless hours in my evenings and weekends.

**Comment by hmic** [comment]
@markstory Please don't give a sh** on my words. You - as the rest of the core and all the members of the community - are doing a great job! I would want to do about anything but not harm the project by discouraging you or others!
